### ISSUE TITLE
2 particulars numbers render the exact same result whatever the salt used ???!!!

### ISSUE THREAD

**Issue by sdespont** [vinkla/hashids#180]
I have a big problem understanding why these 2 particulars numbers render the exact same result whatever the salt used ???!!!

```
        $hashids = new Hashids('any kind of salt', 10);
        $a = $hashids->encode(44888);
        $b = $hashids->encode(44900);
```

Result with this salt **for both numbers** : `egjagwwPBd`

It is always the same result. Is there anyone able to explain why?

I am using PHP 8.0.20 BTW

**Comment by sdespont** [comment]
Really sorry, the numbers are very similar, but different with a letter as lower case.

**Comment by quantumwebco** [comment]
I have this issue, 2 different salts and exactly same hashid for same number

We are using alphabet "ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890" which seems to be the problem, if I add back in the lowercase characters then it makes unique ids

Is this an issue with complexity @vinkla ? Does it need the extra characters in the alphabet to make unique ids?

![image](https://github.com/user-attachments/assets/86e34669-d21d-4e20-a28c-b685f400c1aa)

**Comment by miquelfire** [comment]
@quantumwebco 
> 2 different salts

If you were to compare how this function works compared to an encryption algorithm (Note, Hashids is not Crypto safe, it's basically base64 or something like that. I'm just using encryption as an example), the salt serves the same purpose as the private key. Unless you're including the salt somehow in the output, how are you decoding the results when someone uses the code?

**Comment by quantumwebco** [comment]
We use this trait in our models https://gist.github.com/quantumwebco/ae3a25398960978e4d866c6fe714c8cb

**Comment by miquelfire** [comment]
https://gist.github.com/quantumwebco/ae3a25398960978e4d866c6fe714c8cb#file-hashidtrait-php-L25

If I understand what you're doing, ignore repeats if they're from different classes. You might be using the library for the wrong purpose.

The way you're complaining looks like this, if you were making a SQL database:
"Table A is generating the same primary keys as Table B!"

**Comment by quantumwebco** [comment]
We are using the library to create a 10 digit reference, a bit like an invoice number 

I'm not complaining. I'm simply asking a question why I get the same hash with 2 different salts

If you're not going to offer any help and just gonna be insulting then we'll just stop using the library

**Comment by miquelfire** [comment]
As for how this library works, let's assume the encode function is a complex `toString(intNum, base)` function. In languages that have this feature, if the input number is `10`, then with base set to `2`, you get `1010` as a return value, base `10` is `10`, `7` is `12`, and `16` is `A`. If we just use that `toString()` I gave before as a standalone function, base is the salt. So `toString(16, 16)` would give the same value as `toString(10, 10)` and `toString(2, 2)` and so on.

From what I remembered when I really looked at how the code worked, I think salt acted as either something like `inputNumber xor salt` or the salt acted as a part of the seed (if not the whole seed) for a PRNG type functionally the library uses to shuffle the alphabet.

PS: The closest thing I found to the `toString()` function I talked about that PHP has is [base_convert](https://www.php.net/manual/en/function.base-convert.php)
